apache/airflow
# [Mothership PR] MVP of AIP-69 - A experimental usable version of Edge Executor

This PR is a DRAFT to further get to a MVP release of AIP-69. DO NOT MERGE (yet) - but please feedback on it if you like.

It is a continuation of PR #40224 which was a kind of PoC and step-by-step more stuff came onto this on top.

This is a kind of "Mothership" PR where all contributions are collected. Small parts will be separately opened as PR to get on main and finally I'd split this PR up into smaller pieces for easier review. In this PR all parts and work is collected.

Please treat this as experimental as noted in the docs, it is not expected to be rock-solid but a first usable version. It can be used to distribute the first task. Status of implementation is tracked in `docs/apache-airflow-providers-edge/edge_executor.rst` below headline "Feature Backlog of MVP to Release Readiness" in this PR.

If you want to test it you can:
- Checkout the branch
- Run `breeze start-airflow --load-example-dags --backend postgres --executor EdgeExecutor`
- The Edge Worker will be start in one panel by breeze
- A task can be run, e.g. use example DAG `example_python_operator` or the `Integration Test` with a bit more complexity (this integration test has been moved to `providers/src/airflow/providers/edge/example_dags/integration_test.py`). Wohoo! 
![image](https://github.com/apache/airflow/assets/95105677/c0e148d7-8712-43d4-b927-958d1fd578ad)
- If you open the UI you will see an additional plugin loaded in the menu "Admin" -> "Remote Worker Status": 
![image](https://github.com/apache/airflow/assets/95105677/a7a6ae19-f962-44db-a8c7-cf913a7edbeb)

Based on progress and time I will sometimes re-base and push changes to this branch. But it will not be something that will be review-able.
I will not add any migration logic in case of updates for the additional tables in DB Scheme. If I update the draft PR and DB schema changes, DB might need to be reset. Of course migration support would need to be added before a "real" release.
